java-checks/src/main/java/org/sonar/java/checks/security/HardCodedCredentialsShouldNotBeUsedCheck.java

```diff
@@ -131,90 +149,132 @@ public void visitNode(Tree tree) {
   private void checkArguments(Arguments arguments, CredentialMethod method) {
     for (int targetArgumentIndex : method.indices) {
       ExpressionTree argument = ExpressionUtils.skipParentheses(arguments.get(targetArgumentIndex));
-      if (argument.is(Tree.Kind.STRING_LITERAL, Tree.Kind.NEW_ARRAY)) {
-        reportIssue(argument, ISSUE_MESSAGE);
-      } else if (argument.is(Tree.Kind.IDENTIFIER)) {
-        IdentifierTree identifier = (IdentifierTree) argument;
-        Symbol symbol = identifier.symbol();
-        if (!symbol.isVariableSymbol() || JUtils.isParameter(symbol) || isNonFinalField(symbol) || isReassigned(symbol)) {
-          return;
+      var secondaryLocations = new ArrayList<JavaFileScannerContext.Location>();
+      if (isDerivedFromPlainText(argument, secondaryLocations)) {
+        if (secondaryLocations.isEmpty()) {
+          reportIssue(argument, ISSUE_MESSAGE);
+        } else {
+          reportIssue(argument, ISSUE_MESSAGE, secondaryLocations, null);
         }
+      }
+    }
+  }
 
-        VariableTree variable = (VariableTree) symbol.declaration();
+  private static boolean isDerivedFromPlainText(ExpressionTree expression, List<JavaFileScannerContext.Location> secondaryLocations) {
```

Comment:
Having two different method names for ExpressionTree and other types would be safer when refactoring.